traversal/scanner/GraphIterator.java

```diff
@@ -301,200 +284,146 @@ private boolean verifyLoops() {
             } else if (edge.isRolePlayer()) {
-                Identifier.Variable scope = edge.asRolePlayer().scope();
-                Scopes.Scoped scoped = scopes.getOrInitialise(scope);
-                toIter = edge.asRolePlayer().branchEdge(graphMgr, fromVertex, params).mapSorted(
-                        thingAndRole -> {
-                            scoped.record(edge, thingAndRole.value());
-                            return thingAndRole.key();
-                        },
-                        key -> KeyValue.of(key, null),
-                        ASC
-                );
+                return applyEdgeScope(edge.asRolePlayer().branchEdge(graphMgr, fromVertex, params), edge);
             } else {
-                toIter = edge.branch(graphMgr, fromVertex, params);
-            }
-            if (!edge.to().id().isName() && edge.to().outs().isEmpty() && edge.to().ins().size() == 1) {
-                // TODO: This optimisation can apply to more situations, such as to
-                //       an entire tree, where none of the leaves are referenced by name
-                toIter = toIter.limit(1);
+                return (Forwardable<Vertex<?, ?>, Order.Asc>) edge.branch(graphMgr, fromVertex, params);
             }
+        }
 
-            // TODO: this cast is a tradeoff between losing a lot of type safety in the ProcedureEdge branch() return type vs having a cast
-            return (Forwardable<Vertex<?, ?>, Order.Asc>) toIter;
+        private Forwardable<Vertex<?, ?>, Order.Asc> applyLocalScope(Forwardable<Vertex<?, ?>, Order.Asc> roles) {
+            return roles.filter(role -> {
+                Optional<ProcedureVertex<?, ?>> source = localScope.getRoleVertexSource(role.asThing());
+                return source.isEmpty() || source.get().equals(procedureVertex);
+            }).mapSorted(
+                    role -> {
+                        localScope.record(procedureVertex, role.asThing());
+                        return role;
+                    },
+                    Vertex::asThing,
+                    ASC
+            );
         }
 
+        private Forwardable<Vertex<?, ?>, Order.Asc> applyEdgeScope(Forwardable<KeyValue<ThingVertex, ThingVertex>, Order.Asc> iterator,
+                                                                    ProcedureEdge<?, ?> edge) {
+            Scope scope = scopes.get(edge.asRolePlayer().scope());
+            return iterator.filter(thingAndRole -> {
+                Optional<ProcedureEdge<?, ?>> source = scope.getRoleEdgeSource(thingAndRole.value());
+                return source.isEmpty();
+            }).mapSorted(
+                    thingAndRole -> {
+                        scope.record(edge, thingAndRole.value());
+                        return thingAndRole.key();
+                    },
+                    thing -> KeyValue.of(thing.asThing(), null),
+                    ASC
+            );
+        }
     }
 
-    public static class Scopes {
```

Comment:
got rid of the nested scoping structure, we just have a `Scope` object that corresponds to the roles held by vertex/edges within on Relation.